api/src/auth/drivers/oauth2.ts

```diff
@@ -136,15 +138,31 @@ export class OAuth2AuthDriver extends LocalAuthDriver {
 			throw new InvalidCredentialsException();
 		}
 
+		const userPayload = {
+			provider,
+			first_name: userInfo[this.config.firstNameKey],
+			last_name: userInfo[this.config.lastNameKey],
+			email: email,
+			external_identifier: identifier,
+			role: this.config.defaultRoleId,
+			auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+		};
+
 		const userId = await this.fetchUserId(identifier);
 
 		if (userId) {
-			// Update user refreshToken if provided
-			if (tokenSet.refresh_token) {
-				await this.usersService.updateOne(userId, {
-					auth_data: JSON.stringify({ refreshToken: tokenSet.refresh_token }),
-				});
-			}
+			// Run hook so the end user has the chance to augment the
+			// user that is about to be updated
+			const updatedUserPayload = await emitter.emitFilter(
+				`auth.oauth2.update`,
```

Comment:
Lets do that! I'd much rather have an event that works as expected regardless of auth provider used than a setup that only works for certain conditions 🙂

Reply:
Should be genericable since the provider name is passed. One argument against making it generic is that different provider types may have different signatures e.g. oauth has `accessToken` but LDAP doesn't. I personally have no preference.

Reply:
@aidenfoxx True! The type signature of the actual user payload / access token might be different. I'm thinking we could do a more "expect this to be whatever" structure in the metadata passed to the hook maybe? Something like (pseudo): 

```
{
  driver: "oauth2",
  provider: "github",
  payload: {
    ...whatever we got from github
  }
}
```

Reply:
> Right now userInfo is basically what you describe as payload, but I assume you would also bundle the access token in that?

Yeah I think we can make that a hair more generic and make it a "whatever is returned by/available from the provider" which you can then (asynchronously) turn into the specced user object. That way you can use the oauth token to async ping whatever endpoint of the service you want to pull things like avatars or whatever goodies you want 🙂 

> LDAP is a little special in that sense, since we request the data, payload would always be the same, which is probably fine.

I think that's OK.

Reply:
Can we merge this guys?

Reply:
> Can we merge this guys?

See https://github.com/directus/directus/pull/16812#discussion_r1084276446

Reply:
@Nisthar I plan to integrate the requested changes but couldn't get around to it yet :)

Reply:
> @Nisthar I plan to integrate the requested changes but couldn't get around to it yet :)

```ts
const updatedUserPayload = await emitter.emitFilter(
				`auth.oauth2.update`,
				{},
				{ identifier, provider: this.config.provider, accessToken: tokenSet.access_token, userInfo },
				{ database: getDatabase(), schema: this.schema, accountability: null }
			);
// Update user to update refresh_token and other properties that might have changed
await this.usersService.updateOne(userId, {
	auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
	...updatedUserPayload,
});
```

I am wondering in the code above, can't we let the user(admin) to update the user in the event?

Reply:
Ok nvm, i can see that its already discussed.

Reply:
@rijkvanzanten @aidenfoxx I've implemented the changes as suggested for all drivers where it's possible.

I used `providerPayload` instead of `payload` for the new meta entry, to make it more explicit. Also I didn't destructure it's contents as `userInfo` and `accessToken` could share some keys so it would overwrite, similar issue for `ldap` or if in the future we'd need to add more to content `providerPayload`, it would break existing hook implementations.

Docs are also updated.

Could someone trigger the test pipeline? :)

Thanks!

Reply:
Thanks man, my whole app is dependent on this.